sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala

```diff
@@ -172,6 +172,20 @@ object ParserUtils {
           }
           sb.append(code.asInstanceOf[Char])
           i += 5
+        } else if ((i + 10 < strLength) && b.charAt(i + 1) == 'U' &&
+                   (2 until 10).forall(j => Character.digit(b.charAt(i + j), 16) != -1)) {
+          // \U00000000 style 32-bit unicode character literals.
+
+          val codePoint = Integer.parseInt(b.substring(i + 2, i + 10), 16)
```

Comment:
Actually, a comment: Do you need to treat this as 'unsigned'? if it's more than 2^31 - 1, then it will become negative, and I'm not sure checks below work.

Reply:
Ah, the check should be `0 <= codePoint < 0x10000`.

Reply:
Could be; you can also just use Long.parseLong maybe?

Reply:
I considered to use `Long.parseLong` but is it worth consuming additional 4-byte for `Long` (though the impact should be small) ?

Reply:
For a local var? really won't matter. Whatever's simpler IMHO. I am not sure the math on codePoint below works unless it's a long?

Reply:
O.K, the impact should be small so let's take the simpler one.

Reply:
If we are going to use`regex` to match `\U00000000` pattern, I think `regex` should also be used for other escape characters including `\u0000` and `000`.
So I focus on 32-bit unicode escape in this PR and will refactor the code for all the escape characters with `regex`.

Reply:
OK. It may not be faster. It might be simpler.

Reply:
Sorry the latest comment of mine replies to your comment [here](https://github.com/apache/spark/pull/31096#discussion_r555826852).